array declaration causes panic

The following program `sample.go` triggers a panic:

```go
package main

func main() {
// add a sample
a := [...]int{1, 2, 3, 4, 5, 6, 7, 8}
}
```

Expected result:
```console
$ go run ./sample.go
// ouput
```

Got:
```console
$ yaegi ./sample.go
// ouput
1:33: panic
panic: reflect.Set: value of type [8]int is not assignable to type []int [recovered]
        panic: reflect.Set: value of type [8]int is not assignable to type []int

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc0000a3590, 0xc0000bae00)
        /home/chronos/user/go/src/github.com/containous/yaegi/interp/run.go:107 +0x1e6
panic(0xbf8b20, 0xc000116560)
        /usr/local/share/go/src/runtime/panic.go:522 +0x1b5
reflect.Value.assignTo(0xc06f20, 0xc0000360c0, 0x191, 0xd19dd1, 0xb, 0xbed200, 0x0, 0xbf7a20, 0xc00002a788, 0x82)
        /usr/local/share/go/src/reflect/value.go:2339 +0x42b
reflect.Value.Set(0xbed200, 0xc00000f4e0, 0x197, 0xc06f20, 0xc0000360c0, 0x191)
        /usr/local/share/go/src/reflect/value.go:1473 +0xa8
github.com/containous/yaegi/interp.arrayLit.func1(0xc0000a3590, 0xd42780)
        /home/chronos/user/go/src/github.com/containous/yaegi/interp/run.go:1333 +0x1b4
github.com/containous/yaegi/interp.runCfg(0xc0000bae00, 0xc0000a3590)
        /home/chronos/user/go/src/github.com/containous/yaegi/interp/run.go:112 +0x70
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0000c0d80, 0xc0000bab00, 0x0)
        /home/chronos/user/go/src/github.com/containous/yaegi/interp/run.go:93 +0x129
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0000c0d80, 0xc000028390, 0x26, 0xc000028360, 0x25, 0xd0fa15, 0x1, 0xc000028390)
        /home/chronos/user/go/src/github.com/containous/yaegi/interp/interp.go:291 +0x3ad
github.com/containous/yaegi/interp.(*Interpreter).Repl(0xc0000c0d80, 0xc000010010, 0xc000010018)
        /home/chronos/user/go/src/github.com/containous/yaegi/interp/interp.go:334 +0x1c3
main.main()
        /home/chronos/user/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:85 +0x4f7
```

--- a/_test/a33.go
+++ b/_test/a33.go
@@ -1,0 +1,12 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]int{1, 2, 3}
+	b := a
+	fmt.Println(b)
+}
+
+// Output:
+// [1 2 3]

--- a/_test/a34.go
+++ b/_test/a34.go
@@ -1,0 +1,12 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]int{1, 2, 3}
+	var b [3]int = a
+	fmt.Println(b)
+}
+
+// Output:
+// [1 2 3]

--- a/_test/a35.go
+++ b/_test/a35.go
@@ -1,0 +1,13 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]int{1, 2, 3}
+	b := a
+	b[0] = -1
+	fmt.Println(a)
+}
+
+// Output:
+// [1 2 3]

--- a/_test/a36.go
+++ b/_test/a36.go
@@ -1,0 +1,13 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]int{1, 2, 3}
+	var b [3]int = a
+	b[0] = -1
+	fmt.Println(a)
+}
+
+// Output:
+// [1 2 3]

--- a/_test/a37.go
+++ b/_test/a37.go
@@ -1,0 +1,11 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]int{1, 2, 3}
+	fmt.Println(a)
+}
+
+// Output:
+// [1 2 3]

--- a/_test/a38.go
+++ b/_test/a38.go
@@ -1,0 +1,11 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]byte{}
+	fmt.Printf("%T\n", a)
+}
+
+// Output:
+// [0]uint8

--- a/_test/a39.go
+++ b/_test/a39.go
@@ -1,0 +1,12 @@
+package main
+
+import "fmt"
+
+func main() {
+	a := [...]byte{}
+	b := a
+	fmt.Printf("%T %T\n", a, b)
+}
+
+// Output:
+// [0]uint8 [0]uint8

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1743,14 +1743,16 @@ func compositeGenerator(n *node) (gen bltnGenerator) {
 // array variable it is determined from the value's type, otherwise it is
 // computed from the source definition.
 func arrayTypeLen(n *node) int {
-	if n.typ.size != 0 {
+	if n.typ != nil && n.typ.sizedef {
 		return n.typ.size
 	}
 	max := -1
 	for i, c := range n.child[1:] {
 		r := i
 		if c.kind == keyValueExpr {
-			r = int(c.child[0].rval.Int())
+			if v := c.child[0].rval; v.IsValid() {
+				r = int(c.child[0].rval.Int())
+			}
 		}
 		if r > max {
 			max = r

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -395,11 +395,11 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 						dest.typ = src.typ
 					}
 					if dest.typ.sizedef {
-						dest.typ.size = compositeArrayLen(src)
+						dest.typ.size = arrayTypeLen(src)
 						dest.typ.rtype = nil
 					}
 					if sc.global {
-						// Do not overload existings symbols (defined in GTA) in global scope
+						// Do not overload existing symbols (defined in GTA) in global scope
 						sym, _, _ = sc.lookup(dest.ident)
 					} else {
 						sym = &symbol{index: sc.add(dest.typ), kind: varSym, typ: dest.typ}
@@ -1739,8 +1739,13 @@ func compositeGenerator(n *node) (gen bltnGenerator) {
 	return
 }
 
-// compositeArrayLen return the litteral array length, computed from definition
-func compositeArrayLen(n *node) int {
+// arrayTypeLen returns the node's array length. If the expression is an
+// array variable it is determined from the value's type, otherwise it is
+// computed from the source definition.
+func arrayTypeLen(n *node) int {
+	if n.typ.size != 0 {
+		return n.typ.size
+	}
 	max := -1
 	for i, c := range n.child[1:] {
 		r := i

--- a/interp/run.go
+++ b/interp/run.go
@@ -1351,7 +1351,7 @@ func arrayLit(n *node) {
 	}
 
 	var a reflect.Value
-	if n.typ.size > 0 {
+	if n.typ.sizedef {
 		a, _ = n.typ.zero()
 	} else {
 		a = reflect.MakeSlice(n.typ.TypeOf(), max, max)

--- a/interp/type.go
+++ b/interp/type.go
@@ -123,10 +123,11 @@ type itype struct {
 
 // nodeType returns a type definition for the corresponding AST subtree
 func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
-	var err cfgError
-
 	if n.typ != nil && !n.typ.incomplete {
-		return n.typ, err
+		if n.kind == sliceExpr {
+			n.typ.sizedef = false
+		}
+		return n.typ, nil
 	}
 
 	var t = &itype{node: n, scope: sc}
@@ -141,6 +142,7 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 		}
 	}
 
+	var err cfgError
 	switch n.kind {
 	case addressExpr, starExpr:
 		t.cat = ptrT
@@ -158,7 +160,7 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 				t.size = int(n.child[0].rval.Int())
 			case n.child[0].kind == ellipsisExpr:
 				// [...]T expression
-				t.sizedef = true
+				t.size = arrayTypeLen(n.anc)
 			default:
 				if sym, _, ok := sc.lookup(n.child[0].ident); ok {
 					// Resolve symbol to get size value
@@ -182,6 +184,7 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 			if t.val, err = nodeType(interp, sc, n.child[1]); err != nil {
 				return nil, err
 			}
+			t.sizedef = true
 			t.incomplete = t.incomplete || t.val.incomplete
 		} else {
 			if t.val, err = nodeType(interp, sc, n.child[0]); err != nil {
@@ -803,7 +806,7 @@ func (t *itype) refType(defined map[string]bool) reflect.Type {
 	case aliasT:
 		t.rtype = t.val.refType(defined)
 	case arrayT, variadicT:
-		if t.size > 0 {
+		if t.sizedef {
 			t.rtype = reflect.ArrayOf(t.size, t.val.refType(defined))
 		} else {
 			t.rtype = reflect.SliceOf(t.val.refType(defined))
@@ -866,7 +869,7 @@ func (t *itype) frameType() (r reflect.Type) {
 	case aliasT:
 		r = t.val.frameType()
 	case arrayT, variadicT:
-		if t.size > 0 {
+		if t.sizedef {
 			r = reflect.ArrayOf(t.size, t.val.frameType())
 		} else {
 			r = reflect.SliceOf(t.val.frameType())

